assembly/src/library.rs

```diff
@@ -0,0 +1,434 @@
+        Self {
+            major: self.major,
+            minor: self.minor,
+            patch: self.patch + 1,
+        }
+    }
+}
+
+impl Default for Version {
+    fn default() -> Self {
+        Self::MIN
+    }
+}
+
+impl Serializable for Version {
+    fn write_into(&self, target: &mut ByteWriter) -> Result<(), SerializationError> {
+        target.write_u64(self.major);
+        target.write_u64(self.minor);
+        target.write_u64(self.patch);
+        Ok(())
+    }
+}
+
+impl Deserializable for Version {
+    fn read_from(bytes: &mut ByteReader) -> Result<Self, SerializationError> {
+        let major = bytes.read_u64()?;
+        let minor = bytes.read_u64()?;
+        let patch = bytes.read_u64()?;
+        Ok(Self {
+            major,
+            minor,
+            patch,
+        })
+    }
+}
+
+impl fmt::Display for Version {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}.{}.{}", self.major, self.minor, self.patch)
+    }
+}
+
+impl TryFrom<&str> for Version {
+    type Error = SerializationError;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        let mut tokens = value.split('.').map(|n| {
+            n.parse::<u64>()
+                .map_err(|_| SerializationError::InvalidNumber)
+        });
+        let major = tokens
+            .next()
+            .transpose()?
+            .ok_or(SerializationError::UnexpectedEndOfStream)?;
+        let minor = tokens
+            .next()
+            .transpose()?
+            .ok_or(SerializationError::UnexpectedEndOfStream)?;
+        let patch = tokens
+            .next()
+            .transpose()?
+            .ok_or(SerializationError::UnexpectedEndOfStream)?;
+        Ok(Self {
+            major,
+            minor,
+            patch,
+        })
+    }
+}
+
+// LIBRARY
+// ================================================================================================
+
+/// A library definition that provides AST modules for the compilation process.
+pub trait Library {
+    /// The concrete type used to iterate the modules of the library.
+    type ModuleIterator<'a>: Iterator<Item = &'a Module>
+    where
+        Self: 'a;
+
+    /// Returns the root namespace of this library.
+    fn root_ns(&self) -> &LibraryNamespace;
+
+    /// Returns the version number of this library.
+    fn version(&self) -> &Version;
+
+    /// Iterate the modules available in the library.
+    fn modules(&self) -> Self::ModuleIterator<'_>;
+}
+
+/// A concrete implementation of the Library trait. Contains the minimal attributes of a functional
+/// library.
+///
+/// Implementers of the library trait should use this base type to perform serialization into
+/// `masl` files.
+pub struct LibraryContents {
+    /// Root namespace of the library.
+    pub namespace: LibraryNamespace,
+    /// Version of the library.
+    pub version: Version,
+    /// Available modules.
+    pub modules: Vec<Module>,
+}
```

Comment:
nit: I don't love the name `LibraryContent`. Maybe it should be something like `DefaultLibrary` or something similar?

Also, I'm wondering if we should add another constructor here - something like `read_from_dir()`. This would read the library from a directory with `.masm` files - very similar to what happens in `build.rs` in `stdlib`. But this should probably done in another PR.

Reply:
The implementation attempts to solve the issue of now knowing the available namespaces in the given directory. This way, it will load all `masl` files in a given dir, and the user can just pick a namespace from there.

If we specify a file, that will work as well, but when the user will have to know the full path

Reply:
That is correct. This validation is an attempt to simplify future compilation libraries in a directory. If we follow that, we can just read all directory contents and load these modules, and we are guaranteed to not have duplicates.

It is ofc optional and we can remove this check.